### PR TITLE
Refactor LDA classifier to make it more "observable".

### PR DESCRIPTION
**Pull request by geektoni**
* Removed getter and setters in favour of the `get`/`put` architecture;
* Use `put()` to set weights and bias;
* Correct LDA unit tests;

### CODE REVIEW

File: src/shogun/classifier/LDA.cpp [hunk 1]
```diff
@@ -34,8 +34,8 @@ CLDA::CLDA(
     : CDenseRealDispatch<CLDA, CLinearMachine>(), m_gamma(gamma)
 {
 	init();
-	set_features(traindat);
-	set_labels(trainlab);
+	put("features", dynamic_cast<CFeatures*>(traindat));
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: nooooo
I would really try to avoid using this string based setting of parameters inside c++ code. It is more for swig

File: src/shogun/classifier/LDA.h [hunk 2]
```diff
@@ -158,19 +140,6 @@ class CLDA : public CDenseRealDispatch<CLDA, CLinearMachine>
...
 		}
 
-		/** set features
-		 *
-		 * @param feat features to set
-		 */
-		virtual void set_features(CDotFeatures* feat)
-		{
-			if (feat->get_feature_class() != C_DENSE ||
-				feat->get_feature_type() != F_DREAL)
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.h]: these checks are important, need to move them into train if this is removed

File: src/shogun/classifier/LDA.cpp (hunk 1 above)
Comment (karlnapf) [reply src/shogun/classifier/LDA.cpp]: Sorry those are within class methods.... I didnt realise

Comment (karlnapf) [reply src/shogun/classifier/LDA.cpp]: the dynamic cast is unchecked here

Comment (karlnapf) [reply src/shogun/classifier/LDA.cpp]: if it was for me I would make both labels and features read-only (not puttable), but those are only received through a "train" method. But no need to do that here now

File: tests/unit/classifier/LDA_unittest.cc [hunk 3]
```diff
@@ -200,10 +200,10 @@ TEST(LDA, num_classes_in_labels_exception)
 	SGVector<float64_t> lab{1, -1, 2};
 	SGMatrix<float64_t> feat(1, 3);
 	auto labels = some<CMulticlassLabels>();
-	labels->set_labels(lab);
+	labels->put("labels", lab);
```
Comment (karlnapf) [inline tests/unit/classifier/LDA_unittest.cc]: why not use the ctor above?

File: tests/unit/classifier/LDA_unittest.cc [hunk 4]
```diff
@@ -200,10 +200,10 @@ TEST(LDA, num_classes_in_labels_exception)
...
 	auto features = some<CDenseFeatures<float64_t>>(feat);
 	auto lda = some<CLDA>(0, SVD_LDA);
-	lda->set_labels(labels);
+	lda->put("labels", dynamic_cast<CLabels*>(labels.get()));
```
Comment (karlnapf) [inline tests/unit/classifier/LDA_unittest.cc]: is there the need to use .get? I think there is a put for some, or?

File: src/shogun/classifier/LDA.h (hunk 2 above)
Comment (geektoni) [reply src/shogun/classifier/LDA.h]: LDA uses features dispatching, so there is no need for these checks anymore since we are sure that we will get that type of features.

File: src/shogun/classifier/LDA.cpp (hunk 1 above)
Comment (geektoni) [reply src/shogun/classifier/LDA.cpp]: err I'm a bit confused. Anyway, I just assign them directly now.

File: tests/unit/classifier/LDA_unittest.cc (hunk 4 above)
Comment (geektoni) [reply tests/unit/classifier/LDA_unittest.cc]: Yes, there is a `put` for `Some` indeed.

File: src/shogun/classifier/LDA.cpp [hunk 5]
```diff
@@ -66,7 +71,6 @@ bool CLDA::train_machine_templated(CDenseFeatures<ST>* data)
 {
 	index_t num_feat = data->get_num_features();
 	index_t num_vec = data->get_num_vectors();
-	;
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: :dancer:

File: tests/unit/classifier/LDA_unittest.cc [hunk 6]
```diff
@@ -105,15 +105,15 @@ void test_with_method(
 	generate_test_data<float64_t>(lab, feat);
 
 	auto features = some<CDenseFeatures<float64_t>>(feat);
-	auto labels = some<CBinaryLabels>(lab);
+	auto labels = Some<CLabels>::from_raw(new CBinaryLabels(lab));
```
Comment (karlnapf) [inline tests/unit/classifier/LDA_unittest.cc]: I don't understand why this change is needed. could you explain?

Comment (karlnapf) [reply tests/unit/classifier/LDA_unittest.cc]: ah because of the downcast

Comment (geektoni) [reply tests/unit/classifier/LDA_unittest.cc]: That's because of `put(std::string, Some<>)`. It works only for types which are `is_sg_base`, therefore I need to pass `CLabels`.

Comment (karlnapf) [reply tests/unit/classifier/LDA_unittest.cc]: yep, all good.

**Comment by geektoni** [comment]
This will go again to the `feature/observable_framework` branch instead of `develop`, since I am using the new `observe()` method instead of `put()`.

**Comment by geektoni** [comment]
Merging since MacOS MKL failures are unrelated.
